### PR TITLE
Ensure FxCop in analyzed public API surface for ported FxCop design rules

### PR DESCRIPTION
**Pull request by mavasani**
Majority of FxCop design rules analyze only externally visible members. However, lot of our ported analyzers don't match the FxCop behavior.
See details https://github.com/dotnet/roslyn-analyzers/issues/1432#issuecomment-349031715.

This change fixes the ported analyzers for FxCop design rules to match legacy FxCop.

### CODE REVIEW

File: src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/AbstractTypesShouldNotHaveConstructors.cs [hunk 1]
```diff
@@ -43,7 +43,7 @@ public override void Initialize(AnalysisContext context)
         private static void AnalyzeSymbol(SymbolAnalysisContext context)
         {
             var symbol = context.Symbol as INamedTypeSymbol;
-            if (symbol.IsAbstract)
+            if (symbol.IsAbstract && symbol.GetResultantVisibility() == SymbolVisibility.Public)
```
Comment (333fred) [inline src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/AbstractTypesShouldNotHaveConstructors.cs]: We're doing this often, worth a utility extension `IsPublic()`?

File: src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/AvoidEmptyInterfaces.cs [hunk 2]
```diff
@@ -44,7 +45,11 @@ public override void Initialize(AnalysisContext analysisContext)
         private static void AnalyzeInterface(SymbolAnalysisContext context)
         {
             var symbol = (INamedTypeSymbol)context.Symbol;
-            if (symbol.TypeKind == TypeKind.Interface && !symbol.GetMembers().Any() && !symbol.AllInterfaces.SelectMany(s => s.GetMembers()).Any())
+
+            if (symbol.TypeKind == TypeKind.Interface &&
+                symbol.GetResultantVisibility() == SymbolVisibility.Public &&
+                !symbol.GetMembers().Any() &&
```
Comment (333fred) [inline src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/AvoidEmptyInterfaces.cs]: Not relevant to this PR, but we should really just use Count > 0 here, this is an ImmutableArray. No need for linq allocations.

File: src/Microsoft.CodeQuality.Analyzers/UnitTests/ApiDesignGuidelines/UsePropertiesWhereAppropriateTests.cs [hunk 3]
```diff
@@ -345,6 +376,32 @@ End Class
             GetCA1024BasicResultAt(17, 24, "GetFileNameProtected"));
         }
 
+        [Fact, WorkItem(1432, "https://github.com/dotnet/roslyn-analyzers/issues/1432")]
+        public void VisualBasic_CA1024NoDiagnosticCases_Internal()
+        {
+            VerifyBasic(@"
+Public Class Class1
+	Private fileName As String
```
Comment (333fred) [inline src/Microsoft.CodeQuality.Analyzers/UnitTests/ApiDesignGuidelines/UsePropertiesWhereAppropriateTests.cs]: >	 [](start = 0, length = 1)

This file mixes tabs and spaces. This is horrifying.

**Comment by 333fred** [comment]
I feel like we want to investigate knobs for analyzers. It seems to me that some people will want additional configuration for rules beyond severity, such as for things like this. The PR is fine, but do we have an issue tracking that type of analyzer tuning?

**Comment by jinujoseph** [comment]
Remember speaking to Kasey about it and tuning analyzers is in our Toto list , adding @kuhlenh to track

File: src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/AbstractTypesShouldNotHaveConstructors.cs (hunk 1 above)
Comment (mavasani) [reply src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/AbstractTypesShouldNotHaveConstructors.cs]: Done. I used the name `IsExternallyVisible` to match FxCop terminology.

File: src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/AvoidEmptyInterfaces.cs (hunk 2 above)
Comment (mavasani) [reply src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/AvoidEmptyInterfaces.cs]: Done

File: src/Microsoft.CodeQuality.Analyzers/UnitTests/ApiDesignGuidelines/UsePropertiesWhereAppropriateTests.cs (hunk 3 above)
Comment (mavasani) [reply src/Microsoft.CodeQuality.Analyzers/UnitTests/ApiDesignGuidelines/UsePropertiesWhereAppropriateTests.cs]: Fixed.

**Comment by mavasani** [comment]
> It seems to me that some people will want additional configuration for rules beyond severity

Yes, this has been a constant request from analyzer authors - there needs to be some configuration/options channel from the end user to the analyzers, but needs considerable work:
https://github.com/dotnet/roslyn/issues/3798

**Comment by mavasani** [comment]
@dotnet retest prtest/windows/debug please

**Pull request merged**